File: 2020/day01/src/main.rs
use std::collections::HashSet;
use std::fs;

fn list_of_nums_from_file() -> Vec<i32> {
    let filename = "./input";
    println!("In file {}", filename);
    let content = fs::read_to_string(filename).expect("Something went wrong reading the file");
    content
        .lines()
        .map(|line| line.trim())
        .filter(|line| !line.is_empty())
        .map(|line| line.parse::<i32>().unwrap())
        .collect()
}

fn find_num(nums: Vec<i32>, total: i32) -> Option<i32> {
    let complements: HashSet<i32> = nums.iter().map(|num| total - num).collect();
    for num in nums {
        if complements.contains(&num) {
            return Some(num);
        }
    }
    None
}

#[allow(dead_code)]
fn part_1_solution_1() {
    let nums = list_of_nums_from_file();
    let found = find_num(nums, 2020).unwrap();
    println!("{} {} {}", found, 2020 - found, (found * (2020 - found)));
}

#[allow(dead_code)]
fn part_2_solution_1() {
    let nums = list_of_nums_from_file();
    for num in &nums {
        let sub_nums: Vec<i32> = nums
            .iter()
            .filter(|sub_num| *sub_num != num)
            .cloned()
            .collect();
        let found = find_num(sub_nums, 2020 - num).unwrap();
        if found != 0 {
            println!(
                "{} {} {} {}",
                found,
                2020 - num - found,
                num,
                (found * (2020 - num - found) * num)
            );
            break;
        }
    }
}

fn main() {
    part_2_solution_1();
}

File: 2020/rust/day01/main.py
def list_from_input():
    a = []
    with open("input") as f:
        for l in f.readlines():
            a.append(int(l))
    return a

def find_num(nums, total):
    s = set([total - n for n in nums])
    for num in nums:
        if num in s:
            return num

def part_1():
    a = list_from_input()
    found = find_num(a, 2020)
    print(found, 2020 - found)
    print(found * (2020 - found))

def part_2():
    a = list_from_input()
    for num in a:
        sub_nums = [n for n in a if n != num]
        found = find_num(sub_nums, 2020 - num)
        if found:
            print(found, num, 2020 - num - found)
            print(found * num * (2020 - num - found))
            return

part_1()
part_2()
